File: Code/MGC/MGCDistTransform.m
%% Transform the distance matrices, with column-wise ranking if needed.
%%
%% @param X is a distance matrix;
%% @param Y is a second distance matrix;
%% @param option is a string that specifies which global correlation to build up-on, including 'mgc','dcor','mantel', and 'rank';
%% @param optionRk is a string that specifies whether ranking within column is computed or not.
%%
%% @return A list contains the following output:
%% @return A and B are the centered distance matrices;
%% @return RX and RY are the column rank matrices of X and Y respectively.
%%
%% @export
%% 
function [A,B,RX,RY]=MGCDistTransform(X,Y,option,optionRk)
if nargin<3
    option='mgc'; % default to mgc transform
end
if nargin<4 || strcmp(option,'rank')
    optionRk=1; % do ranking or not, 0 to no ranking
end

% Depending on the choice of the global correlation, properly transform
% each distance matrix
if strcmp(option,'ReducedRank')
    RX=DistRanks(X);
    RY=DistRanks(Y);
    A=RX;
    B=RY;
    n=size(X,1);
    for i=1:n
        [~,aInd]=sort(A(:,i));
        b=sort(B(:,i));
        A(aInd,i)=b;
    end
    option='mgc'; % default to mgc transform
    [A,~]=DistCentering(A,option,0);
    [B,~]=DistCentering(B,option,0);
else
    [A,RX]=DistCentering(X,option,optionRk);
    [B,RY]=DistCentering(Y,option,optionRk);
end

%% An auxiliary function that properly transforms the distance matrix X
%%
%% @param X is a symmetric distance matrix;
%% @param option is a string that specifies which global correlation to build up-on, including 'mgc','dcor','mantel', and 'rank';
%% @param optionRk is a string that specifies whether ranking within column is computed or not.
%%
%% @return A list contains the following output:
%% @return A is the centered distance matrices;
%% @return RX is the column rank matrices of X.
%%
function [A,RX]=DistCentering(X,option,optionRk)
[n]=size(X,1);
if optionRk~=0
    RX=DistRanks(X); % the column ranks for X
else
    RX=zeros(n,n);
end

% If rank transformation, take X as the rank matrix then do default mgc
% transform
if strcmp(option,'rank')
    X=RX;
end

switch option
    case 'dcor' % unbiased dcor transform
        EX=repmat(sum(X,1)/(n-2),n,1)+repmat(sum(X,2)/(n-2),1,n)-sum(sum(X))/(n-1)/(n-2);
    case 'mantel' % mantel transform
        EX=sum(sum(X))/n/(n-1);
    otherwise % Default mgc transform
        EX=repmat(sum(X,1)/(n-1),n,1);
        %     case 'dcor' % original dcor that is biased
        %         EX=repmat(mean(X,1),n,1)+repmat(mean(X,2),1,n)-mean(mean(X));
        %     case 'dcor' % single centering of dcor
        %         EX=repmat(mean(X,1),n,1); % column centering
        %     case 'mcor' % single centering of mcor
        %         EX=repmat(sum(X,1)/n,n,1);
        %         EX=EX+X/n;
end
A=X-EX;

% The diagonal entries are always excluded
for j=1:n
    A(j,j)=0;
end

%% An auxiliary function that sorts the entries within each column by ascending order:
%% For ties, the minimum ranking is used,
%% e.g. if there are repeating distance entries, the order is like 1,2,3,3,4,..,n-1.
%%
%% @param dis is a symmetric distance matrix.
%%
%% @return disRank is the column rank matrices of X.
%%
function [disRank]=DistRanks(dis)

[n,m]=size(dis);
disRank=zeros(n,m);
for i=1:m
    [~,~,a]=unique(dis(:,i));
    disRank(:,i)=a;
end

File: Code/R/MGCDistTransform.R
MGCDistTransform <- function(X,Y,option){
  
  if (missing(option)){
    option='mcor'; # use mcorr by default
  }

  # depending on the choice of the global correlation, properly center the distance matrices
  tA=DistCentering(X,option);
  A=tA$A;
  RX=tA$RX;
  tA=DistCentering(Y,option);
  B=tA$A;
  RY=tA$RX;
  
  if (option!='mcorDouble' && option!='dcorDouble'){
    B=t(B);
    RY=t(RY);
  } 
  
  result=list(A=A,RX=RX,B=B,RY=RY);
  return(result);
}

DistCentering<-function(X,option){
  # An auxiliary function that properly centers the distance matrix X,
  # depending on the choice of global corr.
  n=nrow(X);
  RX=DistRanks(X);
  
  # centering for distance correlation / modified distance correlation /
  # Mantel coefficient
  if (option=='dcor'){ # single centering of dcor
    EX=t(matrix(rep(colMeans(X),n), ncol = n));
  }
  if (option=='mcor'){ # single centering of mcor
    EX=t(matrix(rep(colMeans(X),n), ncol = n));
    EX=EX+X/n;
  }
  if (option=='mgc'){ 
    EX=t(matrix(rep(colMeans(X)*n/(n-1),n), ncol = n));
  }
  if (option=='mantel'){
    EX=sum(X)/n/(n-1);
  }
  if (option=='dcorDouble'){ # original double centering of dcor
    EX=t(matrix(rep(colMeans(X),n), ncol = n))+matrix(rep(rowMeans(X),n), ncol = n)-mean(X);
  }
  if (option=='mcorDouble'){ # original double centering of mcor
    EX=t(matrix(rep(colMeans(X),n), ncol = n))+matrix(rep(rowMeans(X),n), ncol = n)-mean(X);
    EX=EX+X/n;
  }
  A=X-EX;
  
  # for mcor or Mantel, exclude the diagonal entries
  # This is a simpler diagonal modification than the original mcor
  if (option!='mcorDouble' && option!='dcorDouble'){
    for (j in (1:n)){
      A[j,j]=0;
    }
  }
  result=list(A=A,RX=RX);
  return(result);
}

DistRanks <- function(dis) {
  # An auxiliary function that sorts the entries within each column by ascending order.
  #
  # The input is assumed to be a distance matrix.
  #
  # The output is column-wise rank, ordered from 1,...,n.
  #
  # For ties, the minimum ranking is used, 
  # e.g. if there are repeating distance entries, the order is like 1,2,3,3,4,..,n-1.
  n=nrow(dis);
  disRank=matrix(0,n,n);
  for (i in (1:n)){
    v=dis[,i];
    tmp=rank(v,ties.method="min");
    tu=unique(tmp);
    if (length(tu)!=max(tmp)){
      tu=sort(tu);
      for (j in (1:length(tu))){
        kk=which(tmp==tu[j]);
        tmp[kk]=j;
      }
    }
    disRank[,i]=tmp;
  }
  return(disRank);
}